### PR TITLE
Improve the "not exhaustive" diagnostic in the presence of a when clause.

### LINKED ISSUE #45157: Improve the "not exhaustive" diagnostic in the presence of a when clause.
After https://github.com/dotnet/roslyn/pull/44702 we produce a diagnostic that gives an example of a pattern that would cover a non-covered scenario in a switch expression that isn't exhaustive. However, we don't tell the user when that is because of the presence of a `when` clause. @alrz observed this [in code review](https://github.com/dotnet/roslyn/pull/44702#issuecomment-641706686):

> I wonder what example we give for:
> ```cs
> return unsigned switch 
> {
>   0 => true,
>   var x when x > 0 => false,
> };
> ```
> I would expect `_` since any other number is caught by the second arm. F# gives the example anyways but has a special diagnostic in presence of when. `For example, the value '1' may indicate a case not covered by the pattern(s). However, a pattern rule with a 'when' clause might successfully match this value.`

We should definitely enhance the diagnostic to include this information when appropriate.

### PR DESCRIPTION
**Pull request by gafter**
Fixes #45157

### CODE REVIEW

**Comment by gafter** [comment]
@alrz This is based on your suggestion.

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a couple of reviews of this diagnostic improvement?

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a couple of reviews of this diagnostic improvement?

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a second review?

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a second review?

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a second review?

File: src/Compilers/CSharp/Portable/Binder/PatternExplainer.cs [hunk 1]
```diff
@@ -22,23 +22,26 @@ internal static class PatternExplainer
         /// <param name="nodes">The set of nodes in topological order.</param>
         /// <param name="node">The node of interest.</param>
         /// <param name="nullPaths">Whether to permit following paths that test for null.</param>
+        /// <param name="requiresFalseWhenClause">set to true when if the path requires some when clause to evaluate to 'false'</param>
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/PatternExplainer.cs:25]: >when if [](start = 62, length = 7)

Typo?

**Pull request merged**